Skybox

Dhruv-0-Arora opened the pull request:
### Description
Created a spherical skybox using glsl shaders. The skybox is currently implemented in the SceneRenderer.ts file and uses the glsl shaders located in @/shaders. In /public/ there is the custom starry_sky.png used to create the stars. These will most likely be removed or changed.

### Key Features:
- Starry Skybox: Uses custom starry image as the sky for the spherical skybox.
- Color Mixing: Shaders are utilized to blend colors, adding depth and realism. The shaders use the lygia package to incorporate a Perlin noise algorithm which allows for the generation of different patterns.
- Theme Integration Ready: The implementation has the ability to utilize the current theme. However, it currently uses hard-coded colors to create the skybox.

> [!WARNING]
> The renderer for the robot has been commented out in this code. There are also numerous changes to the tsconfig & vite config.


[JIRA Issue](https://jira.autodesk.com/browse/AARD-1668)

Dhruv-0-Arora committed:
Comit

Dhruv-0-Arora committed:
Created new skybox with stars. Still need to fix warping

Dhruv-0-Arora committed:
reduced warping and optimized the code section

HunterBarclay commented on fission/public/skycube_1/skyrender0001.bmp:
Please remove all the texture files. If we do end up using a texture in the future, we'll include it in the assetpack.

HunterBarclay commented on fission/src/shaders/fragment.glsl:
Adjust the formatting on this file to be consistent and bundle code you think is worth saving in large comments at the end of the file.

HunterBarclay replied:
And detail the code you save

BrandonPacewic replied:
@Dhruv-0-Arora Moved the commented out code sections into his own fork on another branch.

HunterBarclay commented on fission/src/systems/scene/SceneRenderer.ts:
```diff
@@ -80,6 +85,38 @@ class SceneRenderer extends WorldSystem {
+        console.log('Current Theme:', currentTheme['Background']['color']['r']);
+        
+        const textureLoader = new THREE.TextureLoader();
+        const cloudTexture = textureLoader.load('./starry_sky.png');
```
Remove cloud texture

HunterBarclay commented on fission/package.json:
```diff
@@ -5,7 +5,7 @@
   "scripts": {
     "dev": "vite",
-    "build": "tsc && vite build",
+    "build": "tsc && vite build --base=/",
```
```suggestion
    "build": "tsc && vite build",
```

HunterBarclay reviewed (changes requested):
Could you remove the stars?

Dhruv-0-Arora committed:
Added all changes to skybox code from PR

Dhruv-0-Arora commented:
Resolved all changes

BrandonPacewic commented on fission/src/shaders/fragment.glsl:
```diff
@@ -0,0 +1,65 @@
+
+	// vec4 imageColor = texture2D(uTexture, adjustedUV);
+	// vec3 colorMix = mix(vec3(0.0, 0.0, 0.0), blueAccent, noiseValue*0.7);
+	// vec3 skyColor = mix(colorMix, imageColor.rgb, imageColor.a);
```
I would prefer to leave large chunks of commented out code in a separate branch if the goal is to reference back to it later but that would be up to @HunterBarclay.

BrandonPacewic commented on fission/src/systems/scene/SceneRenderer.ts:
```diff
@@ -141,6 +172,52 @@ class SceneRenderer extends WorldSystem {
+
+    // public isPowerOf2(value: number): boolean {
+    //     return (value & (value - 1)) === 0;
+    // }
```
Same as previous comment.

I would prefer to leave large chunks of commented out code in a separate branch if the goal is to reference back to it later but that would be up to @HunterBarclay.

BrandonPacewic commented on fission/src/shaders/fragment.glsl:
```diff
@@ -0,0 +1,65 @@
+// some calculations
+float func(float x) {
+	return x*x*(3.0-2.0*x);
+}
```
What kind of calculations? 🤔

BrandonPacewic commented on fission/src/shaders/fragment.glsl:
```diff
@@ -0,0 +1,65 @@
+
+#define PI 3.14159265358979323846264338327
+
+#include "/node_modules/lygia/generative/snoise.glsl" // this is wack
```
I agree, this definitely looks wack. Is this the best way you could find to go about doing this?

BrandonPacewic commented on fission/src/Synthesis.tsx:
```diff
@@ -166,25 +166,25 @@ function Synthesis() {
+            // await (async () => {
+            //     if (!miraAssembly || !(miraAssembly instanceof mirabuf.Assembly)) {
+            //         return;
+            //     }
         
-                const parser = new MirabufParser(miraAssembly);
-                if (parser.maxErrorSeverity >= ParseErrorSeverity.Unimportable) {
-                    console.error(`Assembly Parser produced significant errors for '${miraAssembly.info!.name!}'`);
-                    return;
-                }
+            //     const parser = new MirabufParser(miraAssembly);
+            //     if (parser.maxErrorSeverity >= ParseErrorSeverity.Unimportable) {
+            //         console.error(`Assembly Parser produced significant errors for '${miraAssembly.info!.name!}'`);
+            //         return;
+            //     }
                 
-                const mirabufSceneObject = new MirabufSceneObject(new MirabufInstance(parser));
-                World.SceneRenderer.RegisterSceneObject(mirabufSceneObject);
-            })();
+            //     const mirabufSceneObject = new MirabufSceneObject(new MirabufInstance(parser));
+            //     World.SceneRenderer.RegisterSceneObject(mirabufSceneObject);
+            // })();
```
I see you mentioned that this code was commented out in the PR description, was there a reason for this?

Dhruv-0-Arora replied:
Hunter had commented it so it would run faster on my computer

BrandonPacewic replied:
👍

BrandonPacewic reviewed (commented):
Some minor clarifying points. When you get the chance try to resolve the merge conflicts (unless @HunterBarclay does it for you 🙃). If you need any help with that feel free to grab me!

Dhruv-0-Arora committed:
Adding requested changes to PR
removed redundant variables in fragment & vertex glsl files
removed redundant uTime from code (previously used for moving clouds)
Deleted useful code fragments that were commented out
uncommented mirabuf

Dhruv-0-Arora committed:
Merge remote-tracking branch 'upstream/dev' into dhruv/skybox

BrandonPacewic commented on fission/src/systems/scene/SceneRenderer.ts:
```diff
@@ -68,11 +73,34 @@ class SceneRenderer extends WorldSystem {
+        this._scene.add(ground);
+
+        // skybox
+        const currentTheme = (window as any).getTheme();
```
ESLint wants a different type here instead of `any`. Any chance we can change this?

HunterBarclay replied:
@Dhruv-0-Arora What is this `getTheme()` function?

HunterBarclay replied:
If you're trying to get the background to update to the scene, I would build out a component that, when re-rendered, updates the uniforms of the skybox

Dhruv-0-Arora replied:
the get theme function is supposed to get the currentTheme. Right now its just black

Dhruv-0-Arora replied:
I think I probably need to fix it

Dhruv-0-Arora committed:
Fixed ESLint issue + updated so theme pulls background colors properly

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/skybox

Dhruv-0-Arora committed:
Adding requested changes to skybox to utlize useTheme() method from ThemeContext
avoiding the use of 'any'

Dhruv-0-Arora commented:
@HunterBarclay @BrandonPacewic  do the changes to skybox fix everything for you?

HunterBarclay committed:
Merge branch 'dev' into dhruv/skybox

HunterBarclay reviewed (approved):
Looking good for now. We'll be changing up theme and what not pretty significantly so we'll address changes then

BrandonPacewic commented on fission/src/systems/scene/SceneRenderer.ts:
```diff
@@ -137,6 +161,21 @@ class SceneRenderer extends WorldSystem {
+    /* 
+     * Updates the skybox colors based on the current theme
+
+     * @param currentTheme: current theme from ThemeContext.useTheme()
```
Probably best to have this than not. I would prefer that we standardize whether or not we want these kinds of doc strings at some point but that's a moot point anyways.

Looks good to me!
